src/jdk.incubator.vector/share/classes/jdk/incubator/vector/Float16.java

```diff
@@ -0,0 +1,1667 @@
+        // characters rather than codepoints.
+
+        if (trialResult == 0.0 // handles signed zeros
+            || Math.abs(trialResult) > (65504.0 + 32.0) || // Float.MAX_VALUE + ulp(MAX_VALUE)
+            noDoubleRoundingToFloat16(trialResult)) {
+            return valueOf(trialResult);
+        } else {
+            // If double rounding is not ruled out, re-parse, create a
+            // BigDecimal to hold the exact numerical value, round and
+            // return.
+
+            // Remove any trailing FloatTypeSuffix (f|F|d|D), not
+            // recognized by BigDecimal (or BigInteger)
+            int sLength = s.length();
+            if (Character.isAlphabetic(s.charAt(sLength-1))) {
+                s = s.substring(0, sLength - 1);
+            }
+
+            char startingChar = s.charAt(0);
+            boolean isSigned = (startingChar == '-') || (startingChar == '+');
+            // Hex literal will start "-0x..." or "+0x..." or "0x...""
+            boolean hexInput = isX(s.charAt(isSigned ? 2 : 1));
+
+            if (!hexInput) { // Decimal input
+                // Grammar of BigDecimal string input is compatible
+                // with the decimal grammar for this method after
+                // trimming and removal of any FloatTypeSuffix.
+                return valueOf(new BigDecimal(s));
+            } else {
+                // For hex inputs, convert the significand and
+                // exponent portions separately.
+                //
+                // Rough form of the hex input:
+                // Sign_opt 0x IntHexDigits_opt . FracHexDigits_opt [p|P] SignedInteger
+                //
+                // Partition input into between "0x" and "p" and from
+                // "p" to end of string.
+                //
+                // For the region between x and p, see if there is a
+                // period present. If so, the net exponent will need
+                // to be adjusted by the number of digits to the right
+                // of the (hexa)decimal point.
+                //
+                // Use BigInteger(String, 16) to construct the
+                // significand -- accepts leading sign.
+                StringBuilder hexSignificand = new StringBuilder();
+                if (isSigned) {
+                    hexSignificand.append(startingChar);
+                }
+
+                int fractionDigits = 0;
+                int digitStart = isSigned ? 3 : 2 ;
+
+                int periodIndex = s.indexOf((int)'.');
+
+                int pIndex = findPIndex(s);
+
+                // Gather the significand digits
+                if (periodIndex != -1) {
+                    // Reconstruct integer and fraction digit sequence
+                    // without the period.
+                    hexSignificand.append(s.substring(digitStart,      periodIndex));
+                    hexSignificand.append(s.substring(periodIndex + 1, pIndex));
```

Comment:
```suggestion
                    hexSignificand.append(s, digitStart, periodIndex);
                    hexSignificand.append(s, periodIndex + 1, pIndex);
```

Reply:
Good refactoring; thanks for the suggestion.